doc/specs/#120,121,119-Update-Uninstall-List.md

```diff
@@ -0,0 +1,230 @@
+
+If the installed version is greater than or equal to the version of the app available in the repo, then the "available update" will be left blank.
+
+### Installed out of band from the Windows Package Manaager but not in repo
+If the app is installed out of band from the Package Manager, and is not available in the repo, then the Windows Package Manager will treat it as if there are no updates available.
+
+
+
+
+## Mapping between Windows Package Manager repo and Apps and Features
+*StoreApps*  
+Store apps will be mapped via the The Package Family Name (which is composed by Publisher and Name).   The Package Family Name will be consistent between Apps and Features and the Windows Package Manager repo
+
+*MSIs and EXEs*  
+Will be mapped to the Windows Package Manager repo via the SystemAppId.  The SystemAppId is a required field.  The value in the field differs depending on the InstallerType.
+
+For MSI it is the product code.  Typically a GUID that is typically found in the uninstall registry location and includes the brackets.
+For example: ```{5740BD44-B58D-321A-AFC0-6D3D4556DD6C}```
+
+```HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{3740BD44-B58D-321A-AFC0-6D3D4556DD6C}]```
+    
+```HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall\{3740BD44-B58D-321A-AFC0-6D3D4556DD6C}]```
+
+For inno, wix, nullsoft, and exe, the SystemAppId should be a string that is located in either of the Uninstall keys above. 
+
+### Upgrade
+
+The **upgrade** command is designed to update one or more applications.  The **upgrade** command when executed with no parameters or with --all will update all packages that have updates available.
+
+
+## Usage
+
+`winget upgrade [[-q] \<query>] [\<options>]`
+
+![show command]()
+
+**Upgrade** when executed by itself will upgrade all apps ready for updates by the Windows Package Manager.
+
+
+| Argument  | Description |
+|--------------|-------------|
+| **-q,--query** |  The query used to search for an application. |
+| **-?, --help** |  Gets additional help on this command. |
+
+## Options
+
+The following options are available.
+
+| Option  | Description |
+|--------------|-------------|
+| **--all**         |  update all applications that are installed and have updates available.|
+| **--id**         |  Filter results by ID. |
+| **--name**   |      Filter results by name. |
+| **--moniker**   |  Filter results by application moniker. |
+| **--tag** |     Filter results by tag
+| **--command**  | Filter results by command
+| **-s,--source** |   Find the application using the specified [source](source.md). |
+| **-e,--exact**     | Find the application using exact match. |
+
+
+*Upgrade app when no update is available*  
+If the user attempts to update an app that there is no known update for, the package manager will issue an error.
```

Comment:
Should `winget` exit with a non-successful status code if this is the case?

Reply:
Yes, I believe so.  If the command requested cannot be executed, the process should return a failure exit code.

Reply:
This seems counter intuitive to me, and it's also not what Linux package managers do as far as I know (using `pacman` here). If the purpose of the `upgrade` command is to check if newer version of package(s) are available and get them, then it didn't fail if the last available package(s) are installed already. The output should be different (f.e. "there is nothing to do") but the command would be considered a success anyway.

Reply:
I guess it comes down to what the purpose is deemed to be semantically.  I am not opposed to the purpose being "Make sure I have the latest version of this", which then I agree with you that no error is correct.  I was reading it more as "Get a newer version of this", which while I agree is a subtle difference, it is still the difference between a command that cannot be executed, versus one that can.

I suppose there is the question of what is done in the --all case.  In that case, should there be an error if no updates are available for anything?  That feels wrong to me, as I don't think anyone would ever say "Get a newer version of *everything*".  So for consistency, I believe that the semantics should be Ensure, not Get.

So I agree that there should not be an error exit code, and the verbage here should be "issue a warning" not an error.